### DISCUSSION TITLE
Optimizing yielding prior to actions out of the box

### DISCUSSION THREAD

**Discussion by felixarntz** [WordPress/gutenberg#64729] [Interactivity API]
Over the past few months, the Interactivity API was enhanced with several `data-wp-on-async*` directives (see #61885), which allow invoking actions asynchronously, i.e. yielding to the main thread prior to execution the action, which helps performance by avoiding contention.

First of all, it's worth clarifying that asynchronous actions and the `data-wp-on-async` directives are separate concepts, even though they are somewhat related:
* Asynchronous actions are generator functions, which for instance allows breaking up their logic into distinct tasks by yielding to the main thread.
* Any actions invoked by `data-wp-on-async` directives will additionally yield to the main thread before even starting to run the action. The critical consideration here is that this is only reasonable if the action does not make use of certain synchronous `event` methods, such as `event.preventDefault()` or `event.stopPropagation()`.
* In other words: You can invoke both asynchronous and non-asynchronous actions with either `data-wp-on` or `data-wp-on-async`, as long as the action does not use one of those synchronous event methods. If it does, you must use `data-wp-on`.

The purpose of this discussion is focused mostly on the yielding _prior_ to executing an action, i.e. the functionality enabled by the `data-wp-on-async` directives.

## Current problems
While the `data-wp-on-async` approach is a decent start in preparing the Interactivity API to be more responsible about competing interactions, it currently puts the onus unnecessarily on the developers _consuming_ the store, i.e. the ones that write the HTML containing the directives. It has a few drawbacks:
* The decision on whether an action can be called with `data-wp-on-async` or not purely depends on whether the action does not use one of the synchronous event methods. So it is not a great DX to have to know about whether this is the case for each action as the developer writing the HTML directive.
    * See below code examples to better illustrate this.
* Generally, if an action does _not_ use one of the synchronous event methods, it's always better for performance to call it with `data-wp-on-async` so that it yields to the main thread prior to executing. So this makes it questionable why the person writing the HTML directive even needs to consider this at all.
* Last but not least, the difference between `data-wp-on` and `data-wp-on-async` is complicated to understand for people not terribly familiar with JavaScript. I believe one of the value propositions of the Interactivity API is that in principle it can allow adding interactivity even without writing any JavaScript (if consuming actions from a reusable store from somewhere else, e.g. something core provides, or a 3P package). Having these two directives is IMO unreasonably complex from that perspective, which most likely will lead to a large portion of people using `data-wp-on` in the long run because they don't understand when to use `data-wp-on-async`.
    * This is made worse because the majority of actions does not need to use the synchronous event methods. But because `data-wp-on` will probably be used by more folks (if in doubt), it will lead to a lot of unnecessary synchronous action calls.

## Proposed solution
The overall idea is the following: Whether an action can be called after yielding to the main thread or whether it must be called immediately is purely dependent on the action implementation, so it should be on the developer implementing the action in the store rather than the developer using it.

One important consideration: It is not reasonably possible for the Interactivity API to automatically detect whether an action calls a synchronous event methods or not. Therefore, the developers has to explicitly clarify that themselves.

This leads to the following concrete proposal:
* Introduce a helper function, or a set of helper functions that allow annotating the action implementation with whether it can be run asynchronously (after yielding to the main thread) or not.
* If the action is explicitly annotated in that way (to either allow or disallow asynchronous calling), the Interactivity API should ignore whether the directive is `data-wp-on` or `data-wp-on-async` and instead automatically make the right decision:
    * If the action allows asynchronous calling, always treat it as if `data-wp-on-async` was used.
    * If the action requires synchronous calling, always treat it as if `data-wp-on` was used.
    * If the action is not annotated, handle it like today. This is purely for backward compatibility.
* Short-term to long-term vision after launch:
    1. Promote the use of these functions going forward for every action implementation to establish this as the new best practice.
    2. Further down the line, sing the new functions for implementing actions should become more or less required, i.e. any action not annotated should trigger a deprecation warning.
    3. Finally, the `data-wp-on-async` directives should be deprecated too in favor of all actions being annotated and thus allowing the Interactivity API to make the right decision automatically.

### Proof of concept example
Just a quick example to illustrate what this could look like (let's not get hung up on function names/signatures for now):

**Interactivity API code example:**
```js
function createSyncAction( action ) {
    action.sync = true;
    return action;
}

function createAsyncAction( action ) {
    action.sync = false;
    return action;
}

// Then update directive handlers for all data-wp-on* directives to call `yieldToMain()`
// prior to the action if `action.sync === false`, or to not call it if `action.sync === true`.
```

**Store example:**
```js
import { store, createSyncAction, createAsyncAction } from '@wordpress/interactivity';

const { state, actions } = store( 'test', {
    actions: {
        toggleSomething: createAsyncAction( () => {
            // Logic to contextually toggle an element.
        } ),

        checkFormSubmission: createSyncAction( ( event ) => {
            if ( ! funcThatChecksValidity() ) {
                event.preventDefault();
            }

            // Other logic.
        } ),
    },
    // ...
} );
```

**Directive example:**
```html
<!-- The caller doesn't need to worry about it and can always use data-wp-on. -->
<button data-wp-on--click="actions.toggleSomething" >...</button>
<form data-wp-on--submit="actions.checkFormSubmission" >...</form>
```

I discussed this with @westonruter earlier this week, and we believe such an approach would lead to a much better long-term performance outcome, while simplifying the complexity by directly connecting the decision how to call the action with its implementation and decoupling it from directives.

Curious to get feedback on this proposal and shape this further.

**Comment by felixarntz** [comment]
@luisherranz Related to your comment https://github.com/WordPress/gutenberg/issues/61634#issuecomment-2127851683.

**Reply by luisherranz** [reply]
Hey, this is very interesting, thanks for the ping! 

However, considering that, for backward compatibility, we must still allow people not to use those wrappers…

```js
// This must keep working!
checkFormSubmission: (event) => {
    if (!funcThatChecksValidity()) {
      event.preventDefault();
    }
    // Other logic.
  },
```

…this doesn't solve the problem either "out of the box", does it? We still need _to educate_ people in the same way we need to educate them today to use `data-wp-on-async`.

So this solution simply moves the responsibility of stating that the action doesn't use sync `event` methods from the action's consumer to the creator.

Apart from that:

- I would like to avoid the need to use wrapper functions for all the actions when we have managed to keep the store as a plain JavaScript object so far, which is very clean and simple. It'd be better to use the store property to make those indications, like `actions.checkFormSubmission` and `asyncActions.checkFormSubmission`.
- But, on the other hand, it'd be confusing to call asynchronous actions to the ones that don't need to use `event` synchronous methods when we are also calling asynchronous actions to the ones that need to await promises. So I guess we would need a better naming to make things clearer.

  ```js
  store('..', {
    actions: {
      fetchSomething: createSyncAction(function* (event) {
        // This is a sync action?
        event.preventDefault();
        const { attributes } = getElement();
        yield fetch(attributes.href);
        // ...
      }),
      toggleSomething: createAsyncAction(() => {
        // This is an async action?
        state.isOpen = !state.isOpen;
      }),
    },
  });
  ```

Another possibility we could study is to create a Webpack plugin for `wp-scripts` that does this automatically, transforming this:

```js
store('..', {
  actions: {
    *fetchSomething(event) {
      event.preventDefault();
      const { attributes } = getElement();
      yield fetch(attributes.href);
    },
    toggleSomething() {
      state.isOpen = !state.isOpen;
    },
  },
});
```

Into this:

```js
store('..', {
  actions: {
    *fetchSomething(event) { // event is here, do nothing
      event.preventDefault();
      const { attributes } = getElement();
      yield fetch(attributes.href);
    },
    *toggleSomething() { // no event
      yield splitTask(); // add this
      state.isOpen = !state.isOpen;
    },
  },
});
```

I'm not sure how hard that would be, and it's a solution that would not work when people don't use `wp-scripts`, but maybe it's worth a try. What do you think?

**Reply by DAreRodz** [reply]
I agree that the developer who implements an action should be responsible for determining whether an action should be sync or async, not the developer who consumes it.

In that regard, wouldn't it be enough to call `yield splitTask()` at the top of the function? 🤔

Maybe we could expose a function different than `splitTask()` with a more explicit name. Maybe that could return a value to indicate the action is async, and inform the iAPI whether it is being called under the right directive.

**Reply by DAreRodz** [reply]
@luisherranz, the `toggleSomething()` function you shared in your example above should become a generator function, shouldn't it? ☝️

**Reply by luisherranz** [reply]
Updated, thanks!

**Reply by luisherranz** [reply]
> Maybe we could expose a function different than `splitTask()` with a more explicit name. Maybe that could return a value to indicate the action is async, and inform the iAPI whether it is being called under the right directive.

Could you share an explicit example of this?

**Reply by DAreRodz** [reply]
It would require more thought, haha, but in general, it could work as follows:

1. You call `yield asyncCallback()` (or whatever name we decide to use) at the top of your generator function.
2. That function doesn't need to return a promise. It could be a unique value that the logic evaluating the generator understands. The relevant part of the code is [here](https://github.com/WordPress/gutenberg/blob/bd9ba413cee207338de17cd7d2141debff527074/packages/interactivity/src/utils.ts#L136-L160), inside the `withScope()` util.
3. If that code receives the unique value returned by `asyncAction()` as the result of the first `yield`, then we can mark the function as an _async_ event listener callback, execute `splitTask()` internally, and inform the iAPI that the callback is async.

Using the code you shared before, this is what developers would have to write:

```js
store('..', {
  actions: {
    *fetchSomething(event) {
      event.preventDefault();
      const { attributes } = getElement();
      yield fetch(attributes.href);
    },
    *toggleSomething() {
      yield asyncCallback(); // This is required
      state.isOpen = !state.isOpen;
    },
  },
});
```

It's an explicit API like the one proposed by @felixarntz, but it wouldn't require wrapping the functions with extra calls.

**Reply by cbravobernal** [reply]
And you would be also "helping" devs to learn and understand about generators. We still would need to document "a lot" why and when is needed.

**Reply by luisherranz** [reply]
So, how is that different from or better than asking them to use `yield splitTask()`? 

Also, I don't like the idea that all actions that don't use sync `event` methods have to be generators and include a declaration in the first line. That won't be a good developer experience.

**Reply by felixarntz** [reply]
Thank you for the quick feedback! Some replies inline:
> However, considering that, for backward compatibility, we must still allow people not to use those wrappers…

Yes, of course. But we could still _eventually_ deprecate the way this currently works, i.e. nudge developers more to update the code, while still keeping support for the old way.

> …this doesn't solve the problem either "out of the box", does it? We still need _to educate_ people in the same way we need to educate them today to use `data-wp-on-async`.

It doesn't solve it out of the box when speaking holistically, but it does solve it out of the box from the consumer's perspective (the "consumer" being the person _calling_ the action, not implementing it). It does make a lot more sense to me to have this decision tied to the action itself, not to the directive invoking it.

> So this solution simply moves the responsibility of stating that the action doesn't use sync `event` methods from the action's consumer to the creator.

Exactly, and I think that's a good thing.

> * I would like to avoid the need to use wrapper functions for all the actions when we have managed to keep the store as a plain JavaScript object so far, which is very clean and simple. It'd be better to use the store property to make those indications, like `actions.checkFormSubmission` and `asyncActions.checkFormSubmission`.
> * But, on the other hand, it'd be confusing to call asynchronous actions to the ones that don't need to use `event` synchronous methods when we are also calling asynchronous actions to the ones that need to await promises. So I guess we would need a better naming to make things clearer.

That makes sense. I was hoping y'all would come up with better ideas to implement the solution to this problem, my concrete outline of the functions was just what I came up with. So definitely open to whatever works best. And I completely agree on the name of the functions being confusing with the already existing differentiation between synchronous and asynchronous actions. The intention of what I'm proposing is maybe more understandable as: "Run this action right away" vs "Run this action after yielding to the main thread".

> Another possibility we could study is to create a Webpack plugin for `wp-scripts` that does this automatically, transforming this:
> [...]
>
> I'm not sure how hard that would be, and it's a solution that would not work when people don't use `wp-scripts`, but maybe it's worth a try. What do you think?

I love that idea, definitely worth considering! While it would require to use specific tooling, I think that approach would indeed be a solution that would "magically" work out of the box. Of course we would never be able to guarantee that an action uses it (since not everyone uses `wp-scripts`), so there'd never be a point where we could fully consider the ecosystem "migrated", unlike if this was done at runtime with a deprecation strategy in place. But given how much `wp-scripts` is already promoted and used, I think it's a great idea that would abstract this complexity away from the developers entirely.

> In that regard, wouldn't it be enough to call `yield splitTask()` at the top of the function?

In principle yes, but I also don't think we would want to mandate using async (generator) actions for every single action that doesn't use `event`. If your action does one quick task, it's more intuitive to write it as a regular synchronous function, so then it would be IMO a bit odd to make it a generator just to put `yield splitTask()` on top of it - which would apply to probably the majority of actions.

So I would prefer something along the lines of the Webpack auto-fix idea over asking developers to always do this.

> 1. You call `yield asyncCallback()` (or whatever name we decide to use) at the top of your generator function.
> 2. That function doesn't need to return a promise. It could be a unique value that the logic evaluating the generator understands. The relevant part of the code is [here](https://github.com/WordPress/gutenberg/blob/bd9ba413cee207338de17cd7d2141debff527074/packages/interactivity/src/utils.ts#L136-L160), inside the `withScope()` util.
> 3. If that code receives the unique value returned by `asyncAction()` as the result of the first `yield`, then we can mark the function as an _async_ event listener callback, execute `splitTask()` internally, and inform the iAPI that the callback is async.

This approach could work as well, but it also IMO has drawbacks. It's cleaner in that it doesn't require wrapping the actions in another function, but it would still mean that basically every action not involving `event` would need to become a generator function just for this. Of course for us involved in building the API this makes sense, but I believe a huge portion of the WP developer community is not all that familiar with async actions, and maybe even generators in the first place. So part of what I would like to achieve is to find a solution that also hides some of that complexity away, to give the performance benefits of yielding to pretty much any action, regardless of the developer's JS expertise.

From that perspective, I'm quite intrigued by the Webpack based "auto-fix" approach, which is actually not too far off from this alternative idea. How about going with the `yield asyncCallback()` approach (naming TBD), but also implementing the "auto-fix" variant in Webpack? This would allow developers to implement it themselves, but we would equally support developers that aren't familiar with this and will intuitively prefer to write "regular" synchronous functions.

> Also, I don't like the idea that all actions that don't use sync `event` methods have to be generators and include a declaration in the first line. That won't be a good developer experience.

+1. By doing the Webpack approach, we would solve that, while at the same time not requiring to wrap the action in another function (like in my original `createSyncAction` / `createAsyncAction` proposal).

**Reply by gziolo** [reply]
It looks like a similar problem to what [`withScope`](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/interactivity-api/api-reference.md#withscope) helper tries to address:

> Actions can depend on the scope when they are called, e.g., when you call getContext() or getElement().
>
> When the Interactivity API runtime execute callbacks, the scope is set automatically. However, if you call an action from a callback that is not executed by the runtime, like in a setInterval() callback, you need to ensure that the scope is properly set. Use the withScope() function to ensure the scope is properly set in these cases.
>
> An example, where actions.nextImage would trigger an undefined error without the wrapper:
>
> ```js
> store('mySliderPlugin', {
> 	callbacks: {
> 		initSlideShow: () => {
> 			setInterval(
> 				withScope( () => {
> 					actions.nextImage();
> 				} ),
> 				3_000
> 			);
> 		},
> });
> ```

In effect, the following could be considered:

```js
import { store, withEvent } from '@wordpress/interactivity';

const { state, actions } = store( 'test', {
    actions: {
        toggleSomething() {
            // Logic to contextually toggle an element.
        } ),
        checkFormSubmission: withEvent( (event) => {
            event.preventDefault();

            // Other logic.
        } ),
    },
    // ...
} );
```

`withEvent` is controlled by the runtime, so it could also warn in the case when the developer tries to access the event after it's no longer expected to be accessible when the async part of the action/callback gets executed. So in the example above, `checkFormSubmission` would be sync, but `toggleSomething` would be async out of the box.

By the way, I believe that a similar issue also exists for generators where the first segment is executed today in the sync fashion so devs can safely access the `event` object, but after it moves to the async execution, the same limitation applies as when using `data-wp-on-async`.

Regarding the backward compatibility, in case some existing projects use `event` today without the `withEvent`, then we can always show a deprecation warning by proxying the object and including a prompt to update the code. The webpack approach proposed might be helpful, but it won't be simple to cover all the use possible use cases how dev code callbacks and action. The same issues pop up when linting React code when working with hook or more recently React Compiler. Some examples:

```js

import { checkFormSubmission, checkFormSubmissionHelper } from './utils';

// later
{ 
    checkFormSubmission( event ) {
        event.preventDefault();
        // Other logic.
    } ),
    checkFormSubmission2( { preventDefault } ) {
        preventDefault();
        // Other logic.
    } ),
    checkFormSubmission3: checkFormSubmission ),
    checkFormSubmission4: checkFormSubmissionHelper( additionalParams ) ),
}
```

**Reply by luisherranz** [reply]
I like the `withEvent` API more than the `createSyncAction`, `createAsyncAction` one 🙂

> `withEvent` is controlled by the runtime, so it could also warn in the case when the developer tries to access the event after it's no longer expected to be accessible when the async part of the action/callback gets executed.
> [...]
> I believe that a similar issue also exists for generators where the first segment is executed today in the sync fashion so devs can safely access the event object, but after it moves to the async execution

If the `event` object is only passed as an argument, then there's no problem because the reference is maintained. That is only a problem when getting new references with `getContext` or `getElement`.

For example, this function works perfectly, despite being an asynchronous function, because the reference to the `context` object was obtained before the first `await`, when the scope was correct.

```js
store('...', {
  actions: {
    async someAction() {
      const context = getContext();
      await somethingAsync();
      context.counter; // This is fine.
    },
  },
});
```

The problem arises when the `getContext` function is used after the first `await`, as the scope is no longer correct.

```js
store('...', {
  actions: {
    async someAction() {
      await somethingAsync();
      const context = getContext();
      context.counter; // This will fail.
    },
  },
});
```

This is what we have solved using generators.

Keep in mind that sometimes `getContext` is not explicit, as it might be behind some derived state.

```js
store('...', {
  state: {
    get double() {
      const { counter } = getContext();
      return counter * 2;
    }
  }
  actions: {
    async someAction() {
      await somethingAsync();
      state.double // This will fail.
    },
  },
});
```

In the case of `event`, this problem doesn’t exist because it’s passed as an argument and the reference is always maintained.

```js
store('...', {
  actions: {
    async someAction(event) {
      await somethingAsync();
      event.target.value; // This is fine.
    },
  },
});
```

Of course, if someone tries to use synchronous functions like `event.preventDefault` after the first `await`, they won't work, but that's basic web programming knowledge. I don't think the Interactivity API should get involved in that case.

**Reply by felixarntz** [reply]
@gziolo @luisherranz I like that idea too, and prefer it over the other ones that were discussed so far (although potentially even this idea we could consider "auto-fixing" via Webpack if the developer hasn't fixed it themselves).

To summarize and ensure I got this right: The proposed end goal is that regular actions will always be executed asynchronously. Only actions wrapped with `withEvent()` will be executed synchronously, and only those actions will even receive the `event` object as a parameter. Is that how you envision it?

Regarding deprecation strategy, it makes sense to me what you're describing about showing deprecation warnings by proxying the `event` object. However, how do we handle the actions that haven't been updated yet? While we should deprecate immediately, we still can't immediately _break_ the behavior in question. So there would need to be a transition period of some sort I think. Maybe at first we don't actually change the actions to run async by default, we only introduce `withEvent()` and show deprecation warnings when using `event` without it? A while later, once we consider the transition period long enough, we actually change the behavior so that regular actions will run asynchronously by default and not receive the `event` object. At this point, code not updated will break, but after reasonable time to migrate to the new `withEvent()`. WDYT?

**Reply by luisherranz** [reply]
Yeah, that's what I suggested [here](https://github.com/WordPress/gutenberg/discussions/64729#discussioncomment-10427818) (except that this solution is better because it shifts the responsibility to the action creator):

- Deprecation period: Callbacks receive a proxified `event` object that works well but emits warnings when synchronous methods are used, informing the user that they need to use `withEvent`.
- Breaking change: Callbacks still receive a proxified `event` object, but it no longer works, and when synchronous methods are used, it throws an error informing the user that they need to use `withEvent`.

It works well for me, although I'm not sure if it's compatible with [WordPress's backward compatibility policy](https://developer.wordpress.org/block-editor/contributors/code/backward-compatibility/#how-to-preserve-backward-compatibility-for-a-javascript-function).

**Reply by gziolo** [reply]
> So there would need to be a transition period of some sort I think. Maybe at first we don't actually change the actions to run async by default, we only introduce withEvent() and show deprecation warnings when using event without it? 

Adding `withEvent`, the deprecation on the `event` object, plus updates to the documentation should be the first step. We can decide later when to transition to async first, while in the meantime keep `data-wp-on-async` documented.

**Reply by gziolo** [reply]
@westonruter scanned the plugin directory for Interactivity API usage, and it wasn't that high at the beginning of July: https://wordpress.slack.com/archives/C071CRKGKUP/p1720562665712459 (link requires registration at https://make.wordpress.org/chat/). In the past, we would contact plugin authors if they needed to update their code and help them with the transition, at the time when announcing the deprecation.

**Reply by felixarntz** [reply]
> It works well for me, although I'm not sure if it's compatible with [WordPress's backward compatibility policy](https://developer.wordpress.org/block-editor/contributors/code/backward-compatibility/#how-to-preserve-backward-compatibility-for-a-javascript-function). 

I think with the transition plan this is solid. In a few cases, where the impact is low and the rationale is strong enough, it's okay to make breaking changes like we're discussing here. The API has not seen too much adoption in the ecosystem at this point, so we're still in a better position to make such changes now (than let's say in a year from now). With the deprecation step first and leaving sufficient time until the next step (which only that is technically a breaking change anyway), we're good.

> Adding `withEvent`, the deprecation on the `event` object, plus updates to the documentation should be the first step. We can decide later when to transition to async first, while in the meantime keep `data-wp-on-async` documented.

That sounds great to me.

Trying to put this in a list of steps/phases, to make sure we're aligned. Each phase would be _at least_ several weeks apart.
1. Introduce `withEvent` and trigger deprecations when using `event` without it (via proxied `event` object). This is essentially to prepare the ecosystem for the async first change.
2. Have all actions run async, unless they use `withEvent`. We can consider still passing the proxied `event` object to other functions, but now the synchronous calls would simply fail with an error, so passing it would be mostly for better DX (more clear error messaging where code still does it wrong).
    * At the same time, we also start treating `data-wp-on-async` just the same as `data-wp-on`. In other words, there's no longer a reason to use `data-wp-on-async`.
    * The same applies to the other `data-wp-on*` variants for `window` and `document`.
3. Trigger deprecations warnings when using `data-wp-on-async` and its `window` and `document` equivalents.
    * The documentation on `data-wp-on-async` should be removed (or at least clearly marked as outdated/deprecated).
4. (Optional:) Remove `data-wp-on-async` (and the other async directives) entirely.
    * Marking this as optional, since would not be a big maintenance burden to maintain the deprecated state, and we should probably only remove them if/once we're convinced that usage has pretty much phased out completely.

Does this look right? If everyone agrees, I think the next step would be to get started with code? 🎉

**Reply by gziolo** [reply]
That’s an excellent summary @felixarntz. I agree with the plan 👍

> The documentation on data-wp-on-async should be removed (or at least clearly marked as outdated/deprecated).

Right, that and next step are less important in terms of how far we go with cleanup. The old name should  still be findable in the documentation. In the runtime we should drive the decision taking the bundle size into account.

**Reply by luisherranz** [reply]
The plan sounds good to me. Thanks Felix.

> Each phase would be at least several weeks apart

I don't quite understand this part. Shouldn't we wait for at least a full cycle of WordPress (from version 6.7 to version 6.8, for example) to show the deprecation warning so that developers have time to change their code?

**Reply by luisherranz** [reply]
By the way, is this something you would like to include in WordPress 6.7? Should we add it to the [iteration](https://github.com/WordPress/gutenberg/issues/63232)?

**Reply by felixarntz** [reply]
> I don't quite understand this part. Shouldn't we wait for at least a full cycle of WordPress (from version 6.7 to version 6.8, for example) to show the deprecation warning so that developers have time to change their code?

I'm not sure I understand your question in relation to what I tried to say. 😆 

We don't need to wait before showing deprecation warnings, that can happen right away. What we definitely _have_ to wait with is any breaking changes (e.g. running actions that don't use `withEvent` async by default).

So we would start with introducing `withEvent` and showing deprecation warnings to any action that uses `event` without using `withEvent`. Then we would wait, potentially a WordPress release cycle like you're saying, or more, until we implement the first breaking change. When we do this could also depend on ecosystem adoption.

**Reply by felixarntz** [reply]
> By the way, is this something you would like to include in WordPress 6.7? Should we add it to the [iteration](https://github.com/WordPress/gutenberg/issues/63232)?

Yes, that would be great! Would you like me to open an issue for it, with the plan from above? On another note, is that something one of you would be able to work on in terms of code? Even though I _think_ I have a decent understanding of how the Interactivity API works, I'm not all that comfortable with the code internals of the Interactivity API.

**Reply by luisherranz** [reply]
> I'm not sure I understand your question in relation to what I tried to say. 😆

Sorry, I got confused with the part "several weeks apart", but what you put in the last comment makes perfect sense to me: 

> So we would start with introducing withEvent and showing deprecation warnings to any action that uses event without using withEvent. Then we would wait, potentially a WordPress release cycle like you're saying, or more, until we implement the first breaking change. When we do this could also depend on ecosystem adoption.

**Reply by luisherranz** [reply]
> Would you like me to open an issue for it, with the plan from above?

That would be great, yes.

> is that something one of you would be able to work on in terms of code?

I will be AFK, but maybe @DAreRodz could help with the code, and you could tackle the documentation side? (including reviewing all the files that now show `data-wp-on-async`).

**Reply by felixarntz** [reply]
Sounds like a plan!

I opened https://github.com/WordPress/gutenberg/issues/64944 and https://github.com/WordPress/gutenberg/issues/64946 for now. The first one is what we should focus on for now (and for WP 6.7, if possible).

I'm happy to help where I can, including documentation.

**Reply by felixarntz** [reply]
Should we close this discussion in favor of using the issues going forward?

**Reply by gziolo** [reply]
Thank you so much @felixarntz. Yes, fell free to close it as I don’t see more action items here.

**Reply by westonruter** [reply]
Sorry to be late to the party. I was on vacation when these discussions were happening. I just wanted to draw attention to https://github.com/WordPress/gutenberg/issues/64944#issuecomment-2339280636 which I left on the issue so as to not re-open this discussion (unless warranted). But I'll quote here for convenience:

>> 2\. Have all actions run async, unless they use `withEvent`.
>
> I don't think this is quite right. Even if an action uses the `event` object this doesn't necessitate that it be synchronous. For example, the following actions use the `event` object but they still use `data-wp-on-async` (see examples below). What matters is not whether the `event` object is used but rather whether a synchronous API on the `event` object is used (e.g. `event.preventDefault()`), right? So instead of `withEvent` should it rather be something like `withBlockingCallback`?
> 
> ## Search block
> 
> https://github.com/WordPress/gutenberg/blob/537fb1800ce2784388ec3d130a37bd26f6f2f62f/packages/block-library/src/search/index.php#L194-L195
> 
> https://github.com/WordPress/gutenberg/blob/537fb1800ce2784388ec3d130a37bd26f6f2f62f/packages/block-library/src/search/view.js#L47-L68
> 
> ## Navigation block
> 
> https://github.com/WordPress/gutenberg/blob/537fb1800ce2784388ec3d130a37bd26f6f2f62f/packages/block-library/src/navigation/index.php#L504
> 
> https://github.com/WordPress/gutenberg/blob/537fb1800ce2784388ec3d130a37bd26f6f2f62f/packages/block-library/src/navigation/view.js#L141-L160

Also https://github.com/WordPress/gutenberg/issues/64944#issuecomment-2339293162:

> > 2\. We can consider still passing the proxied `event` object to other functions, but now the synchronous calls would simply fail with an error
> 
> Actually, no error is emitted when `event.preventDefault()` is called. See: https://codepen.io/westonruter/pen/QWXYJQe?editors=1011
> 
> So passing the proxied event should still be done so this warning can be emitted to the console.

Probably best to continue the discussion on #64944, but I wanted to cross-post here for good measure!

**Reply by felixarntz** [reply]
Thanks @westonruter, will reply on the issue.

**Comment by cbravobernal** [comment]
I commented in the [Slack channel](https://wordpress.slack.com/archives/C071CRKGKUP/p1724401166947339?thread_ts=1724364312.508789&cid=C071CRKGKUP), but my thoughts are very similar to Luis's.

 - Developers may confuse `async await` functions  with `createAsyncAction` functions.
 - They will still need to determine the action.
 - "any action not annotated should trigger a deprecation warning" this is vital. I'm afraid most actions won't be annotated. We could include with independence of this approach. Maybe an `eslint` rule?

**Reply by felixarntz** [reply]
> Developers may confuse `async await` functions  with `createAsyncAction` functions.

Definitely, that's fair. I was trying to indicate to not focus too much on the naming yet, I'm sure we could come up with something better in terms of naming and terminology. :)

> They will still need to determine the action.

Can you clarify what you mean by this? Not sure I understand.

> "any action not annotated should trigger a deprecation warning" this is vital. I'm afraid most actions won't be annotated. We could include with independence of this approach. Maybe an `eslint` rule?

To clarify, this was a longer-term plan proposal. I think at this point the Interactivity API is still quite fresh and hasn't seen too much developer adoption yet, so I think we're still at a point where we can pivot things to a certain degree. I definitely didn't mean to suggest we deprecate the "not annotated" functions right away, but maybe more like in let's say 6 months, once Gutenberg and Core have been updated and we have documented the new requirement and spread the word etc.

Anyways, based on the conversations in the above thread, there may be a much better solution in the first place. It's probably worth centering the conversation in there to not get lost in too many threads :)

**Reply by cbravobernal** [reply]
> They will still need to determine the action.

Sorry for not being clear. This means that the developer still needs to know the difference between async actions, asynchronous calls, and when to use them. I'd prefer an approach that prevents overwhelming by automating the selection, like the webpack auto-fix mentioned.

> Anyways, based on the conversations in the above thread, there may be a much better solution in the first place. It's probably worth centering the conversation in there to not get lost in too many threads :)

Yep! Let's keep the discussion there.

**Comment by luisherranz** [comment]
Resharing the approach I shared [here](https://github.com/WordPress/gutenberg/issues/61634#issuecomment-2127851683), as it still might be worth considering:

I had an idea to make `wp-on` synchronous by default. I don't know WordPress's backward compatibility policy, but it is possible to transition `wp-on` from synchronous to asynchronous by informing users over a period of time in the JS console, similar to a deprecation.

It would work as follows. Instead of adding `wp-on-async`, we add `wp-on-sync`. In `wp-on`, we proxy or wrap the event passed to the callback so that the functions `preventDefault`, `stopPropagation`, etc. display a warning in the console urging users to replace `wp-on` with `wp-on-sync`. If we maintain this warning message for several WordPress versions (again, I'm not entirely sure what the policy would be here - three major versions? one year?), we could make the default behavior asynchronous. Then, when someone uses a synchronous method on `wp-on`, an error message would be displayed telling them to change the `wp-on` directive to `wp-on-sync`.

The only downside would be that, until several WordPress versions later, `wp-on` (and `wp-on-sync`) would need to be synchronous.

**Reply by felixarntz** [reply]
> I had an idea to make `wp-on` synchronous by default.

From what I understand, is this a typo and you mean "asynchronous" in this sentence?

This is another potential option, honestly when I first thought about this a few months back, my initial idea was something along those lines, definitely involving proxying the `event` object to emit warnings like you're saying.

The downside of this approach is that it still decouples whether to call the action async or sync from the actual action implementation, which is why at this point I'm not a fan of this approach anymore. Plus, it's a much more "aggressive" approach, with changing the default. Again, that was my initial thought too, but I think the other ideas discussed above are equally promising in terms of adoption and impact with less BC breakage. That said, if we run into other blockers or problems with the alternatives discussed above, I think this would be worth considering given that at least there's clear messages/warnings to the callers. But I am currently most leaning towards the approach of `yield asyncCallback()` and automatically adding it with Webpack if not already there, that is discussed above.

**Reply by luisherranz** [reply]
> The downside of this approach is that it still decouples whether to call the action async or sync from the actual action implementation, which is why at this point I'm not a fan of this approach anymore

I agree 🙂
